Lib/socket.py

```diff
@@ -348,79 +349,86 @@ def makefile(self, mode="r", buffering=None, *,
-                        continue
-                    except OSError as err:
-                        if total_sent == 0:
-                            # We can get here for different reasons, the main
-                            # one being 'file' is not a regular mmap(2)-like
-                            # file, in which case we'll fall back on using
-                            # plain send().
-                            raise _GiveupOnSendfile(err)
-                        raise err from None
-                    else:
-                        if sent == 0:
-                            break  # EOF
-                        offset += sent
-                        total_sent += sent
-                return total_sent
-            finally:
-                if total_sent > 0 and hasattr(file, 'seek'):
-                    file.seek(offset)
-    else:
-        def _sendfile_use_sendfile(self, file, offset=0, count=None):
+        total_sent = 0
+        # localize variable access to minimize overhead
+        selector_select = selector.select
+        try:
+            while True:
+                if timeout and not selector_select(timeout):
+                    raise TimeoutError('timed out')
+                if count:
+                    blocksize = min(count - total_sent, blocksize)
+                    if blocksize <= 0:
+                        break
+                try:
+                    sent = zerocopy_func(fileno, offset, blocksize)
+                except BlockingIOError:
+                    if not timeout:
+                        # Block until the socket is ready to send some
+                        # data; avoids hogging CPU resources.
+                        selector_select()
+                    continue
+                except OSError as err:
+                    if total_sent == 0:
+                        # We can get here for different reasons, the main
+                        # one being 'file' is not a regular mmap(2)-like
+                        # file, in which case we'll fall back on using
+                        # plain send().
+                        raise giveup_err(err)
+                    raise err from None
+                else:
+                    if sent == 0:
+                        break  # EOF
+                    offset += sent
+                    total_sent += sent
+            return total_sent
+        finally:
+            if total_sent > 0 and hasattr(file, 'seek'):
+                file.seek(offset)
+
+    def _sendfile_use_sendfile(self, file, offset=0, count=None):
+        if not (sendfile := getattr(os, 'sendfile', None)):
             raise _GiveupOnSendfile(
                 "os.sendfile() not available on this platform")
+        return self._sendfile_zerocopy(
+            zerocopy_func=partial(sendfile, self.fileno()),
+            giveup_err=_GiveupOnSendfile,
+            file=file,
+            offset=offset,
+            count=count,
+        )
```

Comment:
```suggestion
        return self._sendfile_zerocopy(
            partial(os.sendfile, self.fileno()),
            _GiveupOnSendfile,
            file, offset, count,
        )
```

It's a bit easier to read IMO.